admin/javascript/src/components/form-builder/index.js

```diff
@@ -0,0 +1,229 @@
+import React from 'react';
+import { connect } from 'react-redux';
+import { bindActionCreators } from 'redux';
+import * as schemaActions from 'state/schema/actions';
+import SilverStripeComponent from 'silverstripe-component';
+import FormComponent from 'components/form/index';
+import TextField from 'components/text-field/index';
+import HiddenField from 'components/hidden-field/index';
+import GridField from 'components/grid-field/index';
+import fetch from 'isomorphic-fetch';
+
+import es6promise from 'es6-promise';
+es6promise.polyfill();
+
+// Using this to map field types to components until we implement dependency injection.
+const fakeInjector = {
+
+  /**
+   * Components registered with the fake DI container.
+   */
+  components: {
+    TextField,
+    GridField,
+    HiddenField,
+  },
+
+  /**
+   * Gets the component matching the passed component name.
+   * Used when a component type is provided bt the form schema.
+   *
+   * @param string componentName - The name of the component to get from the injector.
+   *
+   * @return object|null
+   */
+  getComponentByName(componentName) {
+    return this.components[componentName];
+  },
+
+  /**
+   * Default data type to component mappings.
+   * Used as a fallback when no component type is provided in the form schema.
+   *
+   * @param string dataType - The data type provided by the form schema.
+   *
+   * @return object|null
+   */
+  getComponentByDataType(dataType) {
+    switch (dataType) {
+      case 'String':
+        return this.components.TextField;
+      case 'Hidden':
+        return this.components.HiddenField;
+      case 'Text':
+        // Textarea field (not implemented)
+        return null;
+      case 'HTML':
+        // HTML editor field (not implemented)
+        return null;
+      case 'Integer':
+        // Numeric field (not implemented)
+        return null;
+      case 'Decimal':
+        // Numeric field (not implemented)
+        return null;
+      case 'MultiSelect':
+        // Radio field (not implemented)
+        return null;
+      case 'SingleSelect':
+        // Dropdown field (not implemented)
+        return null;
+      case 'Date':
+        // DateTime field (not implemented)
+        return null;
+      case 'DateTime':
+        // DateTime field (not implemented)
+        return null;
+      case 'Time':
+        // DateTime field (not implemented)
+        return null;
+      case 'Boolean':
+        // Checkbox field (not implemented)
+        return null;
+      case 'Custom':
+        return this.components.GridField;
+      default:
+        return null;
+    }
+  },
+};
+
+export class FormBuilderComponent extends SilverStripeComponent {
+
+  constructor(props) {
+    super(props);
+
+    this.formSchemaPromise = null;
+    this.isFetching = false;
```

Comment:
@chillu I get that nothing that doesn't affect the UI should be stored as state, but don't you think it's conceivable that the user might want to render UI state based on `isFetching === true`?